doc/introduction/compiling_circuits.rst

```diff
@@ -0,0 +1,305 @@
+  to left through the circuit and changing the position of the qubits accordingly.
+
+.. note::
+
+    Most compilation transforms support just-in-time compilation with ``jax.jit``.
+
+The :func:`~.pennylane.transforms.compile` transform allows you to chain together
+sequences of quantum function transforms into custom circuit optimization pipelines.
+
+For example, take the following decorated quantum function:
+
+.. code-block:: python
+
+    dev = qml.device('default.qubit', wires=[0, 1, 2])
+
+    @qml.qnode(dev)
+    @qml.compile()
+    def qfunc(x, y, z):
+        qml.Hadamard(wires=0)
+        qml.Hadamard(wires=1)
+        qml.Hadamard(wires=2)
+        qml.RZ(z, wires=2)
+        qml.CNOT(wires=[2, 1])
+        qml.RX(z, wires=0)
+        qml.CNOT(wires=[1, 0])
+        qml.RX(x, wires=0)
+        qml.CNOT(wires=[1, 0])
+        qml.RZ(-z, wires=2)
+        qml.RX(y, wires=2)
+        qml.PauliY(wires=2)
+        qml.CZ(wires=[1, 2])
+        return qml.expval(qml.PauliZ(wires=0))
+
+The default behaviour of :func:`~.pennylane.transforms.compile` is to apply a sequence of three
+transforms: ``commute_controlled``, ``cancel_inverses``, and then ``merge_rotations``.
+
+>>> print(qml.draw(qfunc)(0.2, 0.3, 0.4))
+0: ──H───RX(0.6)──────────────────┤ ⟨Z⟩
+1: ──H──╭X────────────────────╭C──┤
+2: ──H──╰C────────RX(0.3)──Y──╰Z──┤
+
+
+The :func:`~.pennylane.transforms.compile` transform is flexible and accepts a custom pipeline
+of quantum function transforms (you can even write your own!).
+For example, if we wanted to only push single-qubit gates through
+controlled gates and cancel adjacent inverses, we could do:
+
+.. code-block:: python
+
+    from pennylane.transforms import commute_controlled, cancel_inverses
+    pipeline = [commute_controlled, cancel_inverses]
+
+    @qml.qnode(dev)
+    @qml.compile(pipeline=pipeline)
+    def qfunc(x, y, z):
+        qml.Hadamard(wires=0)
+        qml.Hadamard(wires=1)
+        qml.Hadamard(wires=2)
+        qml.RZ(z, wires=2)
+        qml.CNOT(wires=[2, 1])
+        qml.RX(z, wires=0)
+        qml.CNOT(wires=[1, 0])
+        qml.RX(x, wires=0)
+        qml.CNOT(wires=[1, 0])
+        qml.RZ(-z, wires=2)
+        qml.RX(y, wires=2)
+        qml.PauliY(wires=2)
+        qml.CZ(wires=[1, 2])
+        return qml.expval(qml.PauliZ(wires=0))
+
+>>> print(qml.draw(qfunc)(0.2, 0.3, 0.4))
+0: ──H───RX(0.4)──RX(0.2)────────────────────────────┤ ⟨Z⟩
+1: ──H──╭X───────────────────────────────────────╭C──┤
+2: ──H──╰C────────RZ(0.4)──RZ(-0.4)──RX(0.3)──Y──╰Z──┤
+
+.. note::
+
+    The :class:`~.pennylane.Barrier` operator can be used to prevent blocks of code from being merged during
+    compilation.
+
+
+For more details on :func:`~.pennylane.transforms.compile` and the available compilation transforms, visit
+`the compilation documentation
+<https://pennylane.readthedocs.io/en/stable/code/qml_transforms.html#transforms-for-circuit-compilation>`_.
```

Comment:
Unfortunately your second idea does not work, since I want to link a section, and the `qml_transforms.rst` is just an autogenerating stub if I see correctly...

Reply:
PS: I copied the link to the stable webpage from somewhere else, if it is dangerous we should systematically remove this approach